### PR TITLE
handheld mirror traitor item

### PR DESCRIPTION
**Pull request by tristan1333**
## What Does This PR Do
Adds in a new item to the traitor uplink that acts as a portable mirror, allowing all the same customisation options as the ones scattered across the station in a pocket item. The mirror must be held in the hand, the UI disappears if not being held. Currently using a coder sprite for now.

## Why It's Good For The Game
Can be used to play mind games with people, walking right past people who aren't looking for a purple bearded man when an assistant with a pistol just ran through maint.

## Testing
Set self as a traitor, bought item, used item. Used multiple races

## Changelog
:cl:
add: Adds handheld mirror traitor item
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/traitordevices.dm [hunk 1]
```diff
@@ -404,3 +404,22 @@
...
 	times_used = max_uses - 1
 	activate_batterer()
+
+/obj/item/handheld_mirror
+	name = "hand mirror"
+	desc = "Style, on the go!"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "hand_mirror"
+	w_class = WEIGHT_CLASS_TINY
+	var/datum/ui_module/appearance_changer/appearance_changer_holder
```
Comment (GDNgit) [inline code/game/objects/items/devices/traitordevices.dm:414]: this needs to be gc'd properly

File: code/game/objects/items/devices/traitordevices.dm [hunk 2]
```diff
@@ -404,3 +404,22 @@
...
+/datum/datum/ui_module_appearance_changer/appearance_changer_holder/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.hands_state)
+	..(user, ui_key, ui, force_open, master_ui, state)
+
+/obj/item/handheld_mirror/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.hands_state)
+	appearance_changer_holder.ui_interact(user, ui_key, ui, force_open, master_ui, state)
+
+/obj/item/handheld_mirror/attack_self(mob/user)
+	if(ishuman(user))
+		appearance_changer_holder = new(src, user)
+		ui_interact(user)
```
Comment (GDNgit) [inline code/game/objects/items/devices/traitordevices.dm:425]: no newline at the end of the file

Comment (GDNgit) [inline code/game/objects/items/devices/traitordevices.dm:425]:
```suggestion
		ui_interact(user)

/obj/item/handheld_mirror/Destroy()
	QDEL_NULL(appearance_changer_holder)
	return ..()

```
Always gotta null a var'd datum in destroy, or we run into performance issues

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/traitordevices.dm (hunk 1 above)
Comment (tristan1333) [reply code/game/objects/items/devices/traitordevices.dm:414]: GC?

Comment (GDNgit) [reply code/game/objects/items/devices/traitordevices.dm:414]: Garbage collection, essentially making sure things that are deleted get deleted as efficiently as possible as possible

File: code/game/objects/items/devices/traitordevices.dm [hunk 3]
```diff
@@ -404,3 +404,26 @@
...
+
+/obj/item/handheld_mirror
+	name = "hand mirror"
+	desc = "Style, on the go!"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "hand_mirror"
+	w_class = WEIGHT_CLASS_TINY
+	var/datum/ui_module/appearance_changer/appearance_changer_holder
+
+/datum/datum/ui_module_appearance_changer/appearance_changer_holder/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.hands_state)
```
Comment (lewcc) [inline code/game/objects/items/devices/traitordevices.dm:416]: I'm a little confused by what this proc is supposed to be doing, exactly. Regardless, I think you probably don't want the double datum here?
```suggestion
/datum/ui_module_appearance_changer/appearance_changer_holder/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.hands_state)
```

File: code/datums/uplink_items/uplink_general.dm [hunk 4]
```diff
@@ -525,6 +525,13 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	item = /obj/item/storage/box/syndie_kit/safecracking
 	cost = 5
 
+/datum/uplink_item/stealthy_tools/handheld_mirror
+	name = "Hand Held Mirror"
+	desc = "Change and color your hair on the go."
+	reference = "HM"
+	item = /obj/item/handheld_mirror
+	cost = 5
```
Comment (lewcc) [inline code/datums/uplink_items/uplink_general.dm:533]: Will need to be updated as per the big TC inflation PR that was just merged.

File: code/game/objects/items/devices/traitordevices.dm (hunk 3 above)
Comment (tristan1333) [reply code/game/objects/items/devices/traitordevices.dm:416]: That is me being an absolute caveman learning to use tgui, will fix next commit

File: code/datums/uplink_items/uplink_general.dm (hunk 4 above)
Comment (tristan1333) [reply code/datums/uplink_items/uplink_general.dm:533]: That is it scaled up, I originally was putting for 1. I don't think 1 (5 now) is reasonable for the item potential power, but I'm not sure what a comfortable number would be for it atm

File: code/game/objects/items/devices/traitordevices.dm (hunk 3 above)
Comment (lewcc) [reply code/game/objects/items/devices/traitordevices.dm]: Cool, looks a lot better.

**Comment by McRamon** [comment]
you seem to have accidentally edited the toolbox inhand sprite

**Comment by tristan1333** [comment]
> you seem to have accidentally edited the toolbox inhand sprite

oops fixed

File: code/datums/uplink_items/uplink_general.dm [hunk 5]
```diff
@@ -525,6 +525,13 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	item = /obj/item/storage/box/syndie_kit/safecracking
 	cost = 5
 
+/datum/uplink_item/stealthy_tools/handheld_mirror
+	name = "Hand Held Mirror"
+	desc = "Change and color your hair on the go."
```
Comment (Henri215) [inline code/datums/uplink_items/uplink_general.dm:530]: Could describe this item more over here, maybe telling players how small it is and how good it helps with stealth.

**Review by S34NW — APPROVED** [review]
Just needs a deconflict

**Comment by tristan1333** [comment]
> Just a quick question - was the 5 TC cost taken in mind with the inflation costs, or when it was still 20 TC per traitor?

yeah that was in mind of the inflation

**Comment by lewcc** [comment]
Looks like it's still conflicted, I'm afraid

**Comment by Henri215** [comment]
image

**Comment by S34NW** [comment]
You need to merge master my man

**Pull request merged**